src/constbuffer.c

```diff
@@ -394,3 +398,256 @@ IMPLEMENT_MOCKABLE_FUNCTION(, bool, CONSTBUFFER_HANDLE_contain_same, CONSTBUFFER
+                    (void)memcpy(destination + CONSTBUFFER_CONTENT_OFFSET, source->alias.buffer, source->alias.size);
+
+                    /*Codes_SRS_CONSTBUFFER_02_062: [ CONSTBUFFER_to_fixed_size_buffer shall succeed, write in serialized_size how much it used and return CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_OK. ]*/
+                    result = CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_OK;
+                }
+            }
+        }
+    }
+    return result;
+}
+
+CONSTBUFFER_FROM_BUFFER_RESULT CONSTBUFFER_from_buffer(const unsigned char* source, uint32_t size, uint32_t* consumed, CONSTBUFFER_HANDLE* destination)
+{
+    CONSTBUFFER_FROM_BUFFER_RESULT result;
+    if (
+        /*Codes_SRS_CONSTBUFFER_02_063: [ If source is NULL then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG. ]*/
+        (source == NULL) ||
+        /*Codes_SRS_CONSTBUFFER_02_064: [ If consumed is NULL then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG. ]*/
+        (consumed == NULL) ||
+        /*Codes_SRS_CONSTBUFFER_02_065: [ If destination is NULL then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG. ]*/
+        (destination == NULL)
+        )
+    {
+        LogError("invalid arguments const unsigned char* source=%p, uint32_t size=%" PRIu32 ", uint32_t* consumed=%p, CONSTBUFFER_HANDLE* destination=%p", 
+            source, size, consumed, destination);
+        result = CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG;
+    }
+    else
+    {
+        if (size == 0)
+        {
+            /*Codes_SRS_CONSTBUFFER_02_066: [ If size is 0 then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG. ]*/
+            LogError("cannot deserialize from size=%" PRIu32 "", size);
+            result = CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG;
+        }
+        else
+        {
+            uint8_t version;
+            read_uint8_t(source + CONSTBUFFER_VERSION_OFFSET, &version);
+            if (version != CONSTBUFFER_VERSION)
+            {
+                /*Codes_SRS_CONSTBUFFER_02_067: [ If source byte at offset 0 is not 1 (current version) then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_DATA. ]*/
+                LogError("different version (%" PRIu8 ") detected. This module only knows about version %" PRIu8 "", version, CONSTBUFFER_VERSION);
+                result = CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_DATA;
+            }
+            else
+            {
+                /*Codes_SRS_CONSTBUFFER_02_068: [ If source's size is less than sizeof(uint8_t) + sizeof(uint32_t) then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_DATA. ]*/
+                if (size < sizeof(uint8_t) + sizeof(uint32_t))
```

Comment:
>sizeof(uint8_t) + sizeof(uint32_t) [](http://example.com/codeflow?start=27&length=34)

I'd say CONSTBUFFER_CONTENT_OFFSET. Or the sum of the sizes from the format so it is clear why sizeof(uint8_t) + sizeof(uint32_t)

Reply:
I liked this idea very much so I've changed everywhere to use the macros for SIZES.